lib/livebook/runtime/k8s.ex

```diff
@@ -0,0 +1,593 @@
+    %{
+      pod_template: pod_template,
+      docker_tag: docker_tag,
+      home_pvc: home_pvc,
+      namespace: namespace
+    } = config
+
+    manifest =
+      pod_template
+      |> Pod.pod_from_template()
+      |> Pod.add_env_vars([
+        %{"name" => "LIVEBOOK_RUNTIME", "value" => runtime_data},
+        %{
+          "name" => "POD_IP",
+          "valueFrom" => %{"fieldRef" => %{"apiVersion" => "v1", "fieldPath" => "status.podIP"}}
+        },
+        %{
+          "name" => "POD_NAMESPACE",
+          "valueFrom" => %{
+            "fieldRef" => %{"apiVersion" => "v1", "fieldPath" => "metadata.namespace"}
+          }
+        },
+        %{
+          "name" => "POD_NAME",
+          "valueFrom" => %{"fieldRef" => %{"apiVersion" => "v1", "fieldPath" => "metadata.name"}}
+        }
+      ])
+      |> Pod.set_docker_tag(docker_tag)
+      |> Pod.set_home_pvc(home_pvc)
+      |> Pod.set_namespace(namespace)
+      |> Pod.add_container_port(remote_port)
+
+    case Kubereq.create(req, manifest) do
+      {:ok, %{status: 201, body: %{"metadata" => %{"name" => pod_name}}}} ->
+        {:ok, pod_name}
+
+      {:ok, %{body: body}} ->
+        {:error, "could not create Pod, reason: #{body["message"]}"}
+
+      {:error, error} ->
+        {:error, "could not create Pod, reason: #{Exception.message(error)}"}
+    end
+  end
+
+  defp get_cluster_data(nil), do: %{node_base: "k8s_runtime", remote_port: 44444}
+
+  defp get_cluster_data(_kubeconfig) do
+    local_port = get_free_port!()
+    %{node_base: "remote_runtime_#{local_port}", remote_port: 44444, local_port: local_port}
+  end
+
+  defp k8s_forward_port(nil, _, _, _, _), do: :ok
+
+  defp k8s_forward_port(
+         kubeconfig,
+         context,
+         %{local_port: local_port, remote_port: remote_port},
+         pod_name,
+         namespace
+       ) do
+    with {:ok, kubectl_path} <- find_kubectl_executable() do
+      ports = "#{local_port}:#{remote_port}"
+
+      # We want the proxy to accept the same protocol that we are
+      # going to use for distribution
+      bind_addr =
+        if Livebook.Utils.proto_dist() == :inet6_tcp do
+          "[::1]"
+        else
+          "127.0.0.1"
+        end
+
+      args =
+        [
+          "port-forward",
+          "--kubeconfig",
+          Path.expand(kubeconfig),
+          "--context",
+          context,
+          "-n",
+          namespace,
+          pod_name,
+          ports,
+          "--address",
+          bind_addr
+        ]
+
+      port =
+        Port.open(
+          {:spawn_executable, kubectl_path},
+          [:binary, :hide, :stderr_to_stdout, args: args, env: []]
+        )
+
+      port_ref = Port.monitor(port)
+
+      result =
+        receive do
+          {^port, {:data, "Forwarding from " <> _}} ->
+            :ok
+
+          {^port, {:data, "Error: " <> error}} ->
+            {:error, "failed to port-forward. Error: #{String.trim(error)}"}
```

Comment:
@mruoss please double check. I run:

```
$ kubectl port-forward pod/mypod 5000 6000
Error from server (NotFound): pods "mypod" not found
```

and the error has no `:`, so I think we want this:

```suggestion
          {^port, {:data, "Error " <> _ = message}} ->
            {:error, "failed to port-forward. #{String.trim(message)}"}
```

Reply:
oh wow, what a catch! Yes you're absolutely right. I guess I didn't test this path.